nodejs/lib/util/http.ts

```diff
@@ -0,0 +1,236 @@
+ * Therefore all requests will respond to specific
+ * errors by attempting the fallback hosts, and no
+ * assumptions about host or network is retained to
+ * influence the handling of any subsequent request.
+ *
+ ***************************************************/
+
+const handler = function(uri: string, params: unknown, callback: RequestCallback) {
+	callback = callback || noop;
+	return function(err: Error, response: RequestResponse, body: unknown) {
+		if(err) {
+			callback(err);
+			return;
+		}
+		const statusCode = response.statusCode, headers = response.headers;
+		if(statusCode >= 300) {
+			switch(headers['content-type']) {
+				case 'application/json':
+					body = JSON.parse(body as string);
+					break;
+				case 'application/x-msgpack':
+					body = msgpack.decode(body as Buffer);
+			}
+			const error = (body as { error: ErrorInfo }).error ? ErrorInfo.fromValues((body as { error: ErrorInfo }).error) : new ErrorInfo(
+				(headers['x-ably-errormessage'] as string) || 'Error response received from server: ' + statusCode + ' body was: ' + Utils.inspect(body),
+				Number(headers['x-ably-errorcode']),
+				statusCode
+			);
+			callback(error, body, headers, true, statusCode);
+			return;
+		}
+		callback(null, body, headers, false, statusCode);
+	};
+};
+
+function shouldFallback(err: ErrnoException) {
+	const { code, statusCode } = err;
+	return code === 'ENETUNREACH' ||
+		code === 'EHOSTUNREACH'     ||
+		code === 'EHOSTDOWN'        ||
+		code === 'ETIMEDOUT'        ||
+		code === 'ESOCKETTIMEDOUT'  ||
+		code === 'ENOTFOUND'        ||
+		code === 'ECONNRESET'       ||
+		code === 'ECONNREFUSED'     ||
+		(statusCode >= 500 && statusCode <= 504);
+}
+
+
+function getHosts(client: Rest | Realtime): string[] {
+	/* If we're a connected realtime client, try the endpoint we're connected
+		* to first -- but still have fallbacks, being connected is not an absolute
+		* guarantee that a datacenter has free capacity to service REST requests. */
+	const connection = (client as Realtime).connection;
+	const connectionHost = connection && connection.connectionManager.host;
+
+	if(connectionHost) {
+		return [connectionHost].concat(Defaults.getFallbackHosts(client.options));
+	}
+
+	return Defaults.getHosts(client.options);
+}
+
+const Http: typeof IHttp = class {
+	static methods = [HttpMethods.Get, HttpMethods.Delete, HttpMethods.Post, HttpMethods.Put, HttpMethods.Patch];
+	static methodsWithoutBody = [HttpMethods.Get, HttpMethods.Delete];
+	static methodsWithBody = [HttpMethods.Post, HttpMethods.Put, HttpMethods.Patch];
+
+	/* Unlike for doUri, the 'rest' param here is mandatory, as it's used to generate the hosts */
+	static do(method: HttpMethods, rest: Rest, path: PathParameter, headers: Record<string, string> | null, body: unknown, params: RequestParams, callback: RequestCallback): void {
+		const uriFromHost = (typeof(path) == 'function') ? path : function(host: string) { return rest.baseUri(host) + path; };
```

Comment:
```suggestion
		const uriFromHost = (typeof(path) === 'function') ? path : function(host: string) { return rest.baseUri(host) + path; };
```

Also, I'm curious as to why it's `typeof(path)` (function-like) here but `typeof` (operator-like) elsewhere - e.g. `typeof responseText` later on in this file. 🤔 This may well just be a gap in my JS knowledge (something obvious I'm not seeing), but I think it would be nice to be consistent.

Reply:
(no gap in your knowledge, just style differences - I (and Owen) have generally gone for the operator-like style (to make it clear that it isn't actually a function), but Paddy I think liked the function-like style)